src/lib/operators/jit_operator/jit_operations.cpp

```diff
@@ -25,49 +25,92 @@ namespace opossum {
+    return {false, result.is_null};                                              \
+  }
+
+#define JIT_IS_NOT_NULL_CASE(r, types)                                           \
+  case JIT_GET_ENUM_VALUE(0, types): {                                           \
+    const auto result = left_side.compute<JIT_GET_DATA_TYPE(0, types)>(context); \
+    return {false, !result.is_null};                                             \
+  }
+
+JitValue<bool> jit_not(const JitExpression& left_side, JitRuntimeContext& context) {
   // If the input value is computed by a non-jit operator, its data type is int but it can be read as a bool value.
-  DebugAssert(
-      (lhs.data_type() == DataType::Bool || lhs.data_type() == DataType::Int) && result.data_type() == DataType::Bool,
-      "invalid type for jit operation not");
-  result.set<bool>(!lhs.get<bool>(context), context);
-  result.set_is_null(lhs.is_null(context), context);
+  DebugAssert((left_side.result().data_type() == DataType::Bool || left_side.result().data_type() == DataType::Int),
+              "invalid type for jit operation not");
+  const auto value = left_side.compute<bool>(context);
+  return {value.is_null, !value.value};
 }
 
-void jit_and(const JitTupleValue& lhs, const JitTupleValue& rhs, const JitTupleValue& result,
-             JitRuntimeContext& context) {
-  // If the input values are computed by non-jit operators, their data type is int but they can be read as bool values.
+JitValue<bool> jit_and(const JitExpression& left_side, const JitExpression& right_side, JitRuntimeContext& context) {
+  // three-valued logic AND
+
+  // Get result types
+  const auto lhs = left_side.result();
+  const auto rhs = right_side.result();
+
+  // If the input value is computed by a non-jit operator, its data type is int but it can be read as a bool value.
   DebugAssert((lhs.data_type() == DataType::Bool || lhs.data_type() == DataType::Int) &&
-                  (rhs.data_type() == DataType::Bool || rhs.data_type() == DataType::Int) &&
-                  result.data_type() == DataType::Bool,
+                  (rhs.data_type() == DataType::Bool || rhs.data_type() == DataType::Int),
               "invalid type for jit operation and");
 
-  // three-valued logic AND
-  if (lhs.is_null(context)) {
-    result.set<bool>(false, context);
-    result.set_is_null(rhs.is_null(context) || rhs.get<bool>(context), context);
-  } else {
-    result.set<bool>(lhs.get<bool>(context) && rhs.get<bool>(context), context);
-    result.set_is_null(lhs.get<bool>(context) && rhs.is_null(context), context);
+  const auto left_result = left_side.compute<bool>(context);
+  // Computation of right hand side can be pruned if left result is false and not null
+  if (!left_result.value) {                            // Left result is false
+    if (!lhs.is_nullable() || !left_result.is_null) {  // Left result is not null
+      return {false, false};
```

Comment:
The constructor is not intuitive. I would have expected NULL to be the second parameter, but either way, it's not good to read. Do you run into specialization issues even with `std::optional`?